bin/propolis-server/src/lib/migrate/destination.rs

```diff
@@ -24,176 +27,298 @@ use crate::migrate::probes;
-        // want an error encountered during this send to shadow the run error
-        // from the caller.
-        if let Ok(e) = codec::Message::Error(err.clone()).try_into() {
-            let _ = proto.conn.send(e).await;
+) -> Result<impl DestinationProtocol, MigrateError> {
+    let migration_id = migrate_info.migration_id;
+
+    let log = log.new(slog::o!(
+        "migration_id" => migration_id.to_string(),
+        "migrate_role" => "destination",
+        "migrate_src_addr" => migrate_info.src_addr
+    ));
+
+    info!(log, "negotiating migration as destination");
+
+    // Build upgrade request to the source instance
+    // (we do this by hand because it's hidden from the OpenAPI spec)
+    // TODO(#165): https (wss)
+    // TODO: We need to make sure the src_addr is a valid target
+    let src_migrate_url = format!(
+        "ws://{}/instance/migrate/{}/start",
+        migrate_info.src_addr, migration_id,
+    );
+    info!(log, "Begin migration"; "src_migrate_url" => &src_migrate_url);
+    let (mut conn, _) =
+        tokio_tungstenite::connect_async(src_migrate_url).await?;
+
+    let dst_protocols = super::protocol::make_protocol_offer();
+    conn.send(tungstenite::Message::Text(dst_protocols)).await?;
+    let selected = match conn.next().await {
+        Some(Ok(tungstenite::Message::Text(selected_protocol))) => {
+            info!(log, "source negotiated protocol {}", selected_protocol);
+            match super::protocol::select_protocol_from_offer(
+                &selected_protocol,
+            ) {
+                Ok(Some(selected)) => selected,
+                Ok(None) => {
+                    let offered = super::protocol::make_protocol_offer();
+                    error!(log, "source selected protocol not on offer";
+                           "offered" => &offered,
+                           "selected" => &selected_protocol);
+
+                    return Err(MigrateError::NoMatchingProtocol(
+                        selected_protocol,
+                        offered,
+                    ));
+                }
+                Err(e) => {
+                    error!(log, "source selected protocol failed to parse";
+                           "selected" => &selected_protocol);
+
+                    return Err(MigrateError::ProtocolParse(
+                        selected_protocol,
+                        e.to_string(),
+                    ));
+                }
+            }
         }
-        return Err(err);
-    }
+        x => {
+            conn.send(tungstenite::Message::Close(Some(CloseFrame {
+                code: CloseCode::Protocol,
+                reason: "did not respond to version handshake.".into(),
+            })))
+            .await?;
+            error!(
+                log,
+                "source instance failed to negotiate protocol version: {:?}", x
+            );
+            return Err(MigrateError::Initiate);
+        }
+    };
```

Comment:
style nit, take it or leave it: there are two `match`es here, one of which handles whether or not we receive a text message on the websocket stream, and another which handles whether the text message can be parsed successfully. currently, the second match is nested inside of the first one. IMHO, this code would be a lot easier to read if the two matches were in sequence, rather than nested inside of each other, like this: 
```rust
    let selected_protocol = match conn.next().await {
        Some(Ok(tungstenite::Message::Text(selected_protocol))) => selected_protocol,
        x => {
            conn.send(tungstenite::Message::Close(Some(CloseFrame {
                code: CloseCode::Protocol,
                reason: "did not respond to version handshake.".into(),
            })))
            .await?;
            error!(
                log,
                "source instance failed to negotiate protocol version: {:?}", x
            );
            return Err(MigrateError::Initiate);
        }
    };

    info!(log, "source negotiated protocol {}", selected_protocol);
    let selected = match super::protocol::select_protocol_from_offer(
        &selected_protocol,
    ) {
        Ok(Some(selected)) => selected,
        Ok(None) => {
            let offered = super::protocol::make_protocol_offer();
            error!(log, "source selected protocol not on offer";
                "offered" => &offered,
                "selected" => &selected_protocol);

            return Err(MigrateError::NoMatchingProtocol(
                selected_protocol,
                offered,
            ));
        }
        Err(e) => {
            error!(log, "source selected protocol failed to parse";
                "selected" => &selected_protocol);

            return Err(MigrateError::ProtocolParse(
                selected_protocol,
                e.to_string(),
            ));
        }
    };
```
(I couldn't leave this as a "suggestion" because it contains deleted lines apparently :/)

This way, we are saying "first, try to receive a text message, or bail out if we get something weird. then, try to parse that message". I think that's a bit easier to follow for the reader.